src/vos/vos_io.c

```diff
@@ -1532,9 +1532,11 @@ vos_fetch_begin(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 	if (rc != 0) {
 		daos_recx_ep_list_free(ioc->ic_recx_lists, ioc->ic_iod_nr);
 		ioc->ic_recx_lists = NULL;
-		return vos_fetch_end(vos_ioc2ioh(ioc), NULL, rc);
+		rc = vos_fetch_end(vos_ioc2ioh(ioc), NULL, rc);
+		*ioh = DAOS_HDL_INVAL;
```

Comment:
It is still risky to set and later invalidate the output handle if you experience an error later along the way. It is too easy to overlook another case in the future.

Wouldn't it be simpler to set it only when you are 100% done and nothing can fail?
- replace all `goto set_ioc` would be replaced with `goto out`,
- the `set_ioc` label would be removed
- `*ioh = vos_ioc2ioh(ioc);` would be moved after the last `if`.

So you can exit wherever you would like and the `ioh` would be set only if everything else succeeds. I believe it would be more future-proof.

Please consider my suggestion. Other than that your solution is correct.

Reply:
Your suggestion is reasonable. I will refresh the patch. Hope we still can catch 2.4 timeline.